# **Assessment Test 2**
## **Machine Learning**
### Eivind Johansen

---

---

### **Oppgave A**

In [1]:
import pandas as pd

df_original = pd.read_csv("StudentsPerformance.csv")
df_preprep = df_original.copy()

cols = ["gender", "parental level of education", "test preparation course", "math score", "reading score", "writing score"]
df_preprep = df_preprep[cols]

df_preprep


gender parental level of education test preparation course  math score  \
0    female           bachelor's degree                    none          72   
1    female                some college               completed          69   
2    female             master's degree                    none          90   
3      male          associate's degree                    none          47   
4      male                some college                    none          76   
..      ...                         ...                     ...         ...   
995  female             master's degree               completed          88   
996    male                 high school                    none          62   
997  female                 high school               completed          59   
998  female                some college               completed          68   
999  female                some college                    none          77   

     reading score  writing score  
0               72             74  
1               90             88  
2               95             93  
3               57             44  
4               78             75  
..             ...            ...  
995             99             95  
996             55             55  
997             71             65  
998             78             77  
999             86             86  

[1000 rows x 6 columns]

### **Oppgave B**

In [2]:
def sex_column(gender):
    if gender == "female":
        return 1
    else:
        return 0

df_preprep["sex"] = df_preprep["gender"].apply(sex_column)

del df_preprep["gender"]

df_preprep

In [3]:
def preparation_course_column(test_preparation_course):
    if test_preparation_course == "completed":
        return 1
    else:
        return 0

df_preprep["preparation course"] = df_preprep["test preparation course"].apply(preparation_course_column)

del df_preprep["test preparation course"]

df_preprep


parental level of education  math score  reading score  writing score  \
0             bachelor's degree          72             72             74   
1                  some college          69             90             88   
2               master's degree          90             95             93   
3            associate's degree          47             57             44   
4                  some college          76             78             75   
..                          ...         ...            ...            ...   
995             master's degree          88             99             95   
996                 high school          62             55             55   
997                 high school          59             71             65   
998                some college          68             78             77   
999                some college          77             86             86   

     sex  preparation course  
0      1                   0  
1      1                   1  
2      1                   0  
3      0                   0  
4      0                   0  
..   ...                 ...  
995    1                   1  
996    0                   0  
997    1                   1  
998    1                   1  
999    1                   0  

[1000 rows x 6 columns]

### **Oppgave C**

In [4]:
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder(sparse=False)
transformed = ohe.fit_transform(df_preprep[["parental level of education"]])
df_preprep[ohe.categories_[0]] = transformed

del df_preprep["parental level of education"]

df_preprep


math score  reading score  writing score  sex  preparation course  \
0            72             72             74    1                   0   
1            69             90             88    1                   1   
2            90             95             93    1                   0   
3            47             57             44    0                   0   
4            76             78             75    0                   0   
..          ...            ...            ...  ...                 ...   
995          88             99             95    1                   1   
996          62             55             55    0                   0   
997          59             71             65    1                   1   
998          68             78             77    1                   1   
999          77             86             86    1                   0   

     associate's degree  bachelor's degree  high school  master's degree  \
0                   0.0                1.0          0.0              0.0   
1                   0.0                0.0          0.0              0.0   
2                   0.0                0.0          0.0              1.0   
3                   1.0                0.0          0.0              0.0   
4                   0.0                0.0          0.0              0.0   
..                  ...                ...          ...              ...   
995                 0.0                0.0          0.0              1.0   
996                 0.0                0.0          1.0              0.0   
997                 0.0                0.0          1.0              0.0   
998                 0.0                0.0          0.0              0.0   
999                 0.0                0.0          0.0              0.0   

     some college  some high school  
0             0.0               0.0  
1             1.0               0.0  
2             0.0               0.0  
3             0.0               0.0  
4             1.0               0.0  
..            ...               ...  
995           0.0               0.0  
996           0.0               0.0  
997           0.0               0.0  
998           1.0               0.0  
999           1.0               0.0  

[1000 rows x 11 columns]

### **Oppgave D**

In [5]:
# Her vet vi at vi har et tak på score, så her kan vi bruke MinMaxScaler:

from sklearn.preprocessing import MinMaxScaler

reading_score_scaler = MinMaxScaler()
writing_score_scaler = MinMaxScaler()

scaled_reading_score = reading_score_scaler.fit_transform(df_preprep[["reading score"]])
scaled_writing_score = writing_score_scaler.fit_transform(df_preprep[["writing score"]])

df_preprep["scaled_reading_score"] = scaled_reading_score
df_preprep["scaled_writing_score"] = scaled_writing_score

del df_preprep["reading score"]
del df_preprep["writing score"]

df_preprep

math score  sex  preparation course  associate's degree  \
0            72    1                   0                 0.0   
1            69    1                   1                 0.0   
2            90    1                   0                 0.0   
3            47    0                   0                 1.0   
4            76    0                   0                 0.0   
..          ...  ...                 ...                 ...   
995          88    1                   1                 0.0   
996          62    0                   0                 0.0   
997          59    1                   1                 0.0   
998          68    1                   1                 0.0   
999          77    1                   0                 0.0   

     bachelor's degree  high school  master's degree  some college  \
0                  1.0          0.0              0.0           0.0   
1                  0.0          0.0              0.0           1.0   
2                  0.0          0.0              1.0           0.0   
3                  0.0          0.0              0.0           0.0   
4                  0.0          0.0              0.0           1.0   
..                 ...          ...              ...           ...   
995                0.0          0.0              1.0           0.0   
996                0.0          1.0              0.0           0.0   
997                0.0          1.0              0.0           0.0   
998                0.0          0.0              0.0           1.0   
999                0.0          0.0              0.0           1.0   

     some high school  scaled_reading_score  scaled_writing_score  
0                 0.0              0.662651              0.711111  
1                 0.0              0.879518              0.866667  
2                 0.0              0.939759              0.922222  
3                 0.0              0.481928              0.377778  
4                 0.0              0.734940              0.722222  
..                ...                   ...                   ...  
995               0.0              0.987952              0.944444  
996               0.0              0.457831              0.500000  
997               0.0              0.650602              0.611111  
998               0.0              0.734940              0.744444  
999               0.0              0.831325              0.844444  

[1000 rows x 11 columns]

### **Oppgave E**

In [6]:
import numpy as np

X = np.c_[df_preprep.drop("math score", axis=1)]
y = np.c_[df_preprep["math score"]]

### **Oppgave F**

In [7]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### **Oppgave G**

In [8]:
# I dette tilfellet bruker vi LinerRegression. Vi skal predikere en score og ikke 
# klassifisere om studenten får en score eller ikke.

from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

### **Oppgave H**

In [10]:
from statistics import mean
from sklearn.metrics import mean_absolute_error,mean_squared_error

y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)
y_pred_naiv = np.zeros(len(y_pred_test)) + df_preprep["math score"].mean()

mae_train = mean_absolute_error(y_pred_train, y_train)
mae_test = mean_absolute_error(y_pred_test, y_test)
mae_naiv = mean_absolute_error(y_pred_naiv, y_test)

mse_train = mean_squared_error(y_pred_train, y_train)
mse_test = mean_squared_error(y_pred_test, y_test)
mse_naiv = mean_squared_error(y_pred_naiv, y_test)

rmse_train = np.sqrt(mse_train)
rmse_test = np.sqrt(mse_test)
rmse_naiv = np.sqrt(mse_naiv)

print("Mean Absolute Error:")
print(f"Train: {mae_train}\nTest: {mae_test}\nNaiv: {mae_naiv}\n")
print("Mean Squared Error:")
print(f"Train: {mse_train}\nTest: {mse_test}\nNaiv: {mse_naiv}\n")
print("Root Mean Squared Error:")
print(f"Train: {rmse_train}\nTest: {rmse_test}\nNaiv: {rmse_naiv}")

Mean Absolute Error:
Train: 4.70576171875
Test: 4.3741015625
Naiv: 12.32356

Mean Squared Error:
Train: 34.36610824584961
Test: 30.173871154785157
Naiv: 245.992041

Root Mean Squared Error:
Train: 5.86226135939448
Test: 5.493074836080895
Naiv: 15.684133415652903


Modellen gjør det veldig mye bedre enn å bare predikere gjennomsnittet hver gang. Modellen bommer i gjennomsnitt med 5,49 på test-data'en mens hvis vi bare hadde predikert gjennomsnittet ville vi i snitt bommet med 15.68. Train- og Test score ligger nærme hverandre noe som ikke tyder på at vi har et tilfelle av overfitting. <br> **Konklusjon**: En veldig god modell å bruke.

### **Oppgave H**

I dette tilfellet skal vi predikere om studenten enten har forkurs eller ikke. I slike tilfeller bruker vi DecisionTreeClassifier.

En overfitted modell vil gi et veldig mye bedre resultat enn det vi får ved å bruke test_datene.
Hvis vi ser på "confusion matrix" vil denne for eksempel se slik ut:

**Treningsdata:**
'           | positive | negative |
------------| -------- | -------- |
**positive**    | 94       | 1        |
**negative**    | 0        | 5        |

**Testdata:**
'           | positive | negative |
------------| -------- | -------- |
**positive**    | 65       | 20       |
**negative**    | 10       | 5        |

Her ser vi at resultatet til treningsdataen er mye bedre enn resultatet til testdataen. Derfor tyder dette på at modellen er overfitted. 

Hvis modellen er overfitted vil det si at modellen er for kompleks og vi vil derfor skru ned max_depth for å gjøre modellen mindre kompleks. Hvis modellen er under-fitted vil vi øke kompleksiteten til modellen og derfor øker vi max_depth. 

Vi kan finne ut den optimale løsningen ved å justere max_depth til det punktet der resultatet til det punktet der treningsdataen og testdataen begynner å sprike fra hverandre.